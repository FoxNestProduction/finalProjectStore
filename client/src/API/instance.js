import axios from 'axios';

let store;

export const injectStore = (_store) => {
  store = _store;
};

export const instance = axios.create({
  baseURL: process.env.REACT_APP_API_URL,
});

instance.interceptors.request.use((config) => {
  if (store.getState().authorization && store.getState().authorization.token) {
    // eslint-disable-next-line no-param-reassign
    config.headers.authorization = store.getState().authorization.token;
  }
  return config;
}, (error) => {
  return Promise.reject(error);
});
