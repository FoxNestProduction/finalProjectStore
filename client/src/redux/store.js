import { configureStore, getDefaultMiddleware, combineReducers } from '@reduxjs/toolkit';
// todo: правила для redux-persist
// eslint-disable-next-line import/no-extraneous-dependencies
import storage from 'redux-persist/lib/storage';
// eslint-disable-next-line import/no-extraneous-dependencies
import { persistReducer } from 'redux-persist';
import modalSlice from './slices/modalSlice';
import productsSlice from './slices/productsSlice';
import authorizationSlice from './slices/authorizationSlice';
import userSlice from './slices/userSlice';
import partnersSlice from './slices/partnersSlice';
import searchSlice from './slices/searchSlice';
import errorSlice from './slices/errorSlice';
import reviewsSlice from './slices/reviewsSlice';
import favouriteSlice from './slices/favouriteSlice';
import cartSlice from './slices/cartSlice';
import orderSlice from './slices/orderSlice';
import filterSlice from './slices/filterSlice';
import scrollAnchorSlice from './slices/scrollAnchorSlice';
import newGoogleUserSlice from './slices/newGoogleUserSlice';

const authPersistConfig = {
  key: 'authorization',
  version: 1,
  storage,
};

const userPersistConfig = {
  key: 'user',
  version: 1,
  storage,
};

const favouritePersistConfig = {
  key: 'favorites',
  version: 1,
  storage,
};

const cartPersistConfig = {
  key: 'cart',
  version: 1,
  storage,
};

const reducer = combineReducers({
  error: errorSlice,
  modal: modalSlice,
  partners: partnersSlice,
  authorization: persistReducer(authPersistConfig, authorizationSlice),
  user: persistReducer(userPersistConfig, userSlice),
  newGoogleUser: newGoogleUserSlice,
  products: productsSlice,
  search: searchSlice,
  order: orderSlice,
  reviews: reviewsSlice,
  favourites: persistReducer(favouritePersistConfig, favouriteSlice),
  filter: filterSlice,
  cart: persistReducer(cartPersistConfig, cartSlice),
  scrollAnchor: scrollAnchorSlice,
});

const middleware = getDefaultMiddleware({
  serializableCheck: false,
});

const store = configureStore({
  reducer,
  middleware,
});

export default store;
