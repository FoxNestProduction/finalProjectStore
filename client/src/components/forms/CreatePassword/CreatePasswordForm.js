import React from 'react';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import { Form, Formik } from 'formik';
import { Button } from '@mui/material';
import { useDispatch, useSelector } from 'react-redux';
import LockIcon from '@mui/icons-material/Lock';
import validationSchema from './validationSchema';
import {
  flexcenter,
  container,
  mainTitle,
  inputsWrapper,
  signUpBtn,
} from './styles';
import Input from '../../inputs/Input/Input';
import { closeModal } from '../../../redux/slices/modalSlice';
import { setAuthorization, setToken } from '../../../redux/slices/authorizationSlice';
import { setUser } from '../../../redux/slices/userSlice';
import { setRegistrationError } from '../../../redux/slices/errorSlice';
import { removeDataFromSessionStorage } from '../../../utils/sessionStorageHelpers';
import { CHECKOUT_SS_KEY } from '../../../constants/constants';
import saveUserInfoToSessionStorage from '../../../utils/saveUserInfoToSessionStorage';
import { instance } from '../../../API/instance';
import { fetchCart } from '../../../redux/slices/cartSlice';

export const initialValues = {
  password: '',
};

const CreatePasswordForm = () => {
  const dispatch = useDispatch();
  const newGoogleUser = useSelector((state) => state.newGoogleUser.newGoogleUser);

  const handleSubmit = async (values, actions) => {
    const login = newGoogleUser.email.split('@')[0];
    const newCustomer = {
      ...newGoogleUser,
      ...values,
      login,
      isAdmin: false,
    };
    console.log(newCustomer);

    try {
      const response = await instance.post('/customers', newCustomer);
      const { user, token } = response.data;

      dispatch(setToken(token));
      dispatch(setAuthorization(true));
      dispatch(setUser(user));
      dispatch(closeModal());
      dispatch(setRegistrationError(''));

      removeDataFromSessionStorage(CHECKOUT_SS_KEY);
      saveUserInfoToSessionStorage(user);
      dispatch(fetchCart());
    } catch (error) {
      dispatch(setRegistrationError(error.response.data.message));
      console.error('Помилка реєстрації:', error);
    }
  };

  return (
    <Box
      component="section"
      sx={{
        ...flexcenter,
        ...container,
      }}
    >
      <Typography
        component="h1"
        variant="h2"
        sx={mainTitle}
      >
        Create your own password
      </Typography>
      <Formik
        initialValues={initialValues}
        onSubmit={handleSubmit}
        validationSchema={validationSchema}
      >
        {({ isValid }) => (
          <Form>
            <Box
              sx={{
                ...flexcenter,
                flexDirection: 'column',
                ...inputsWrapper,
              }}
            >
              <Input
                type="password"
                name="password"
                id="registerPassword"
                label="Password"
                placeholder="Сome up with a password"
                icon={<LockIcon />}
              />
              <Button
                disableRipple
                variant="contained"
                sx={signUpBtn}
                type="submit"
                disabled={!isValid}
              >
                Sign up
              </Button>
            </Box>
          </Form>
        )}
      </Formik>
    </Box>
  );
};

export default CreatePasswordForm;
