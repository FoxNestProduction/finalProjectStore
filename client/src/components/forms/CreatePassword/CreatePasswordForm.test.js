import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { Provider, useDispatch } from 'react-redux';
import store from '../../../redux/store';
import CreatePasswordForm from './CreatePasswordForm';
import { instance } from '../../../API/instance';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('../../../API/instance', () => ({
  instance: {
    post: jest.fn(),
  },
}));

describe('Create password form component', () => {
  test('Create password form snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <CreatePasswordForm />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('try to submit empty form', async () => {
    const dispatch = jest.fn();
    useDispatch.mockReturnValue(dispatch);
    instance.post.mockRejectedValue({ response: { data: { message: 'Password is required' } } });
    render(
      <Provider store={store}>
        <CreatePasswordForm />
      </Provider>,
    );

    const submitButton = screen.getByText('Sign up');

    fireEvent.submit(submitButton);

    await waitFor(() => {
      expect(screen.getByText('Password is required')).toBeInTheDocument();
    });
  });

  test('Submit correct form', async () => {
    const dispatch = jest.fn();
    useDispatch.mockReturnValue(dispatch);
    instance.post.mockResolvedValue({ status: 200 });

    render(
      <Provider store={store}>
        <CreatePasswordForm />
      </Provider>,
    );

    const passwordInput = screen.getByPlaceholderText('Ð¡ome up with a password');
    const submitButton = screen.getByText('Sign up');

    fireEvent.change(passwordInput, { target: { value: 'Password12#' } });

    fireEvent.submit(submitButton);

    await waitFor(() => {
      expect(dispatch).toHaveBeenCalled();
    });
  });
});
