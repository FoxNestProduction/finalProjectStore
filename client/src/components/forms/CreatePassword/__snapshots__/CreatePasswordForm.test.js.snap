// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create password form component Create password form snapshot 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root css-1qhxn6p"
  >
    <h1
      class="MuiTypography-root MuiTypography-h2 css-1tzbn0g-MuiTypography-root"
    >
      Create your own password
    </h1>
    <form
      action="#"
    >
      <div
        class="MuiBox-root css-1cj4gsi"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="registerPassword"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-1rmfxk3-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1c8d5n4-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="LockIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
                />
              </svg>
            </div>
            <input
              aria-describedby="helper-text"
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-152mnda-MuiInputBase-input-MuiOutlinedInput-input"
              id="registerPassword"
              name="password"
              placeholder="Сome up with a password"
              type="password"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <button
                aria-label="toggle password visibility"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1afvcqy-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="VisibilityIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
            id="helper-text"
          />
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1d5t28q-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Sign up
        </button>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
